### PR TITLE
AOTAutograd bugfix: outputs can require grad even when compiled in no_grad region

### PR DESCRIPTION
**Pull request by bdhirsh**
Fixes https://github.com/pytorch/pytorch/issues/114338

I used the example in the linked issue as a repro:
```
def fn(x):
    with torch.enable_grad():
        out = x + 1
    return out
    

opt_fn = torch.compile(fn, backend="inductor")  # also, aot_eager, aot_eager_decomp_partition, but not eager
with torch.no_grad():
    res = fn(torch.zeros(10, requires_grad=True))
    opt_res = opt_fn(torch.zeros(10, requires_grad=True))

    assert res.requires_grad == opt_res.requires_grad  # True != False
```

There were two problems:

(1) we would (incorrectly) mark our graph as an inference graph, since no_grad was enabled when we started compiling. This is wrong, because our graph returns outputs that do indeed require_grad (since we created them under an enable_grad)

(2) Even after ensuring that we compile to a training graph, the example above still fails. AOTAutograd compiles a training graph down to a custom autograd.Function, that we invoke at runtime. The compiled code is invoked under a no_grad region however, and so the outputs of the autograd.Function will all not require grad.

I updated the runtime logic to always invoke our custom `autograd.Function` under `enable_grad`, and then manually `.detach` any outputs that we detected at trace time are supposed to not require grad.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115294

### CODE REVIEW

**Review by Chillee — APPROVED** [review]

**Review by zou3519 — APPROVED** [review]

**Review by tugsbayasgalan — APPROVED** [review]

**Comment by bdhirsh** [comment]
Probably going to drop this PR in favor of https://github.com/pytorch/pytorch/pull/114343, since I didn't see that PR the first time around.


Also, I realized the failing tests are because the `.detach` calls I added are unnecessary. We can rely on `ctx.set_non_differentiable` to properly determine which outputs don't require grad
